### PR TITLE
Allow SELECT queries to execute on time-bounded dataflows

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR introduces an `until: Antichain<T>` member to `DataflowDescription`, which indicates through which time the dataflow should run. In particular, this is meant to be useful for dataflows that service `Peek` commands, which are only meant to execute at a single time. These dataflows should properly hydrate at that time, but will produce no further updates and (except for `persist_source`) drop their capabilities and shut down.

This is a bit of a draft at the moment, but it seemed to run ok locally. I have not yet exercised it as compared to `main` to observe any difference. It is possible that it will be nominal.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/compute-client/src/command.proto [hunk 1]
```diff
@@ -84,6 +84,7 @@ message ProtoDataflowDescription {
     repeated ProtoIndexExport index_exports = 4;
     repeated ProtoSinkExport sink_exports = 5;
     optional mz_persist.gen.persist.ProtoU64Antichain as_of = 6;
+    mz_persist.gen.persist.ProtoU64Antichain until = 9;
```
[the remark this answers is not in the document]
Comment (frankmcsherry) [reply src/compute-client/src/command.proto:87]: I'd be delighted for both of them to be non-optional! :D  The "absent" form of `until` being the empty frontier is great for us, as that happens to be the "safest" and also what all prior instances implicitly used.

[the remark this answers is not in the document]
Comment (antiguru) [reply src/compute-client/src/command.proto:87]: Oh for sure, but you're at conflicts with protobuf! Everything is optional, and adding the `optional` keyword only changes the Rust implementation to not create a default value for you if there's none on the wire.

**Comment by vmarcos** [comment]
In a setup with all PK and FK indexes and the hardware as documented in a separate [Notion doc](https://www.notion.so/materialize/TPC-H-Optimization-and-Dataflow-performance-d43aa9cb379d46ccaa0be07e83d658db), we get the following single-worker query processing time for TPC-H queries at scale factor 1 (data in tables):

```
TPC-H Query	At main	With the change
Q1	11152.093 ms (00:11.152)	11196.333 ms (00:11.196)
Q2	74.906 ms	78.897 ms
Q3	530.648 ms	526.452 ms
Q4	2898.169 ms (00:02.898)	2919.613 ms (00:02.920)
Q5	344.711 ms	339.481 ms
Q6	1615.424 ms (00:01.615)	1628.780 ms (00:01.629)
Q7	9693.177 ms (00:09.693)	9605.992 ms (00:09.606)
Q8	255.024 ms	250.510 ms
Q9	4725.561 ms (00:04.726)	1916.131 ms (00:01.916)
Q10	1465.362 ms (00:01.465)	1488.286 ms (00:01.488)
Q11	1347.076 ms (00:01.347)	1380.752 ms (00:01.381)
Q12	2257.867 ms (00:02.258)	2261.439 ms (00:02.261)
Q13	3897.135 ms (00:03.897)	4022.741 ms (00:04.023)
Q14	2035.136 ms (00:02.035)	1957.861 ms (00:01.958)
Q15	1445.697 ms (00:01.446)	1455.743 ms (00:01.456)
Q16	1023.125 ms (00:01.023)	1047.481 ms (00:01.047)
Q17	5998.137 ms (00:05.998)	6011.354 ms (00:06.011)
Q18	16457.729 ms (00:16.458)	16636.207 ms (00:16.636)
Q19	2191.927 ms (00:02.192)	1971.253 ms (00:01.971)
Q20	3461.126 ms (00:03.461)	3393.689 ms (00:03.394)
Q21	8150.921 ms (00:08.151)	8045.009 ms (00:08.045)
Q22	1129.680 ms (00:01.130)	1138.715 ms (00:01.139)
```
I also tried the following query:

```
SELECT COUNT(*) 
FROM (
  SELECT l_returnflag, l_suppkey,
         count(DISTINCT l_linestatus) as ls_count
  FROM lineitem
  GROUP BY l_returnflag, l_suppkey
  LIMIT 10
);
```

```
At main	With the change
3944.948 ms (00:03.945)	3962.567 ms (00:03.963)

```
So right now, there is still no impact, but this may change when we address https://github.com/MaterializeInc/materialize/issues/14351?

The PR itself looks good to me!

**Comment by antiguru** [comment]
Not sure, but it seems there's something going on here:
```
Q9           4725.561 ms (00:04.726)   1916.131 ms (00:01.916)
```

That's .. dramatically better than before?

**Comment by vmarcos** [comment]
> Not sure, but it seems there's something going on here:
> 
> ```
> Q9           4725.561 ms (00:04.726)   1916.131 ms (00:01.916)
> ```
> 
> That's .. dramatically better than before?

Good catch! I tried to reproduce, but the code on main gives me a measurement of `1857.111 ms (00:01.857)` for Q9 individually and of `1863.546 ms (00:01.864)` in a 22-query run after warm-up. The measurements I have in the Notion doc also align with the latter, so I probably just caught a fluctuation in my last run before adding the numbers in.

**Comment by frankmcsherry** [comment]
Quick expectation management on the PR:
1. It's about a. pruning work we don't need to do and 2. allowing clean-up to occur as soon as possible.
2. In TPCH there aren't any changes, so I think no work gets pruned. In a longer running query, we would prune the work associated with changes once we've started, which has the potential to be significant if the query is expensive and the data are changing a lot.
3. It's not clear cleaning up eagerly is the best for latency. It can keep memory low for e.g. `TopK` chains, but also it does work on the critical path that could be deferred until after the query results are ready. It may require multiple workers to elicit, as with a single worker we'll just hit each operator in sequence and shut them down only in a subsequent pass.

**Comment by frankmcsherry** [comment]
I guess also this would be hugely important for `SELECT .. AS OF <old time>` as in the current code base we will load up from the `as_of`, but the entire history from that point onward.

**Comment by frankmcsherry** [comment]
Marked as "ready for review" but no urgency to merge. If it seems actually uncontroversial, might be good to.

**Comment by vmarcos** [comment]
All good points regarding the potential impacts; I think this change definitely moves us forward and unlocks some very interesting possibilities to improve query processing of one-off SELECTs even further. So my vote is to go ahead and merge.

Just out of curiosity, regarding your point 3. above, I ran the TPC-H queries with 8 workers in my setup and I cannot see any big latency impact, not at that scale anyway:

```
At main (ms)    With the change (ms)
1658.169        1695.019
35.946          32.521
118.191         115.085
531.688         514.473
62.411          53.199
242.609         232.129
1729.956        1651.744
54.101          50.313
307.86          297.898
443.986         449.017
239.339         241.581
364.194         374.443
791.73          827.619
333.981         275.907
266.772         242.371
210.219         203.987
868.889         881.685
2795.008        2902.032
372.932         279.164
2764.049        2752.387
1518.873        1451.839
211.24          208.455
```
The query with limit also did not get really affected:

```
At main (ms)       With the change (ms)
606.087            598.593
```

**Review by vmarcos — APPROVED** [review]

**Review by antiguru — APPROVED** [review]

**Comment by frankmcsherry** [comment]
> The query with limit also did not get really affected:

I'm not sure which query this is, but a `limit 10` on a `SELECT` query will not use a `TopK` dataflow. You would need to have a `limit 10` as an internal part of the query (the logic around this is .. hard to explain). And even still, if you just look at latency it may do nothing good and even something bad (depending on the relative costs of allocation and deallocation).

IIrc, one might be able to get something out of 
```sql
select * from (select * from lineitems order by <field> limit 10) order by <other field) limit 10
```
The outer `limit 10` will be stripped off, and the inner `limit 10` will be turned in to dataflow.

**Comment by frankmcsherry** [comment]
A caveat I'll need to figure out: temporal filters can introduce updates at times, even if the inputs are at earlier times. So, probably also need to thread `until` through their evaluation as well. Will investigate and see how large a mess this is.

**Comment by vmarcos** [comment]
> > The query with limit also did not get really affected:
> 
> I'm not sure which query this is, but a `limit 10` on a `SELECT` query will not use a `TopK` dataflow. You would need to have a `limit 10` as an internal part of the query (the logic around this is .. hard to explain). And even still, if you just look at latency it may do nothing good and even something bad (depending on the relative costs of allocation and deallocation).
> 
> IIrc, one might be able to get something out of
> 
> ```sql
> select * from (select * from lineitems order by <field> limit 10) order by <other field>) limit 10
> ```
> 
> The outer `limit 10` will be stripped off, and the inner `limit 10` will be turned in to dataflow.

Apologies for not being so clear; the query I had run before was:

```
SELECT COUNT(*) 
FROM (
  SELECT l_returnflag, l_suppkey,
         count(DISTINCT l_linestatus) as ls_count
  FROM lineitem
  GROUP BY l_returnflag, l_suppkey
  LIMIT 10
);
```

It had a `TopK` operator, but it did not include an `ORDER BY` clause, which made it way cheaper to process than the pattern you proposed. So I instantiated your pattern as:

`
select * from (select * from lineitem order by l_suppkey limit 10) order by l_returnflag limit 10;
`

Then, I got the following measurements:

```
Workers   At main (ms)   With the change (ms)
1         76204.52       72119.671
8         19932.677      18313.789 

```

The query processing times here were a bit less stable than with TPC-H queries, so I repeated the query a few times to get some confidence that the above numbers would be fairly representative. The change seems to have a positive impact, with the query processing time prior to the change being a factor of 1.09x larger at 8 workers than the time after the change.

File: src/compute/src/render/context.rs [hunk 2]
```diff
@@ -519,7 +524,7 @@ where
             let key = key_val.map(|(k, _v)| k);
             return self.as_specific_collection(key.as_deref());
         }
-        let (stream, errors) = self.flat_map(key_val, || {
+        let (stream, errors) = self.flat_map(key_val, move || {
```
Comment (frankmcsherry) [inline src/compute/src/render/context.rs:527]: This is unnecessary.

**Pull request merged**